### ISSUE TITLE
help needed, how to store a activerecord object in an attribute

### ISSUE THREAD

**Issue by Roupiye** [leastbad/all_futures#10]
i tried using Marshal to serialize a obj into an attribute, but the dump method return will be saved as nil instead of a string, and it will throw a error when trying to load the nil value(it will expect a mashal string)

![image](https://user-images.githubusercontent.com/38011914/155340435-b061d8d2-ad5e-4502-80fe-d85d00efa908.png)

![image](https://user-images.githubusercontent.com/38011914/155340481-4d37cf49-0091-4ca1-b276-63baefd9f05e.png)

**Comment by leastbad** [comment]
Hi Breno!

I want to help. I'd love to see you get your solution working. It would help me if you could give a little bit more context for why you are doing this eg what is the desired outcome that is different from what the library already offers?

The first red flag is that the `binary` type is not one of the supported types. You can learn more in https://allfutures.leastbad.com/usage/attributes including a bit about the `binary` issues.

Second, you're using a feature - `serialize` - that isn't technically advertised as being on offer. That said, technically Active Entity *does* support `serialize` based on https://github.com/jasl/activeentity#others which tells me that you should **not** be `include`ing `ActiveModel::Serialization` as it might conflict with something @jasl did to build the library - many features have custom code and it pays to ask "should I even do this" before making assumptions and losing time.

Now, All Futures has no `has_one`, `has_many` or `belongs_to` macros. I have thought about adding macros for `has_one` and `has_many` but they would just do something like this:

```ruby
class PostDraft < AllFutures::Base
  attribute :post_id, :integer
  attribute :tag_ids, :integer, array: true, default: []
  has_one :post, :post_id
  has_many :tags, :tag_ids, where: {category: "colors"}
end
```

However, I haven't implemented this, yet. If it would help, I will expedite. Note: there will never, ever be `belongs_to` support in Active Record that can work with All Futures. It would be a cold day in hell...

Long story short: there's currently an issue with `binary` types in All Futures that I'm 90% sure is actually an issue with Active Entity. We can ask @jasl to look into it, but there's no promises.

In the meantime, I would recommend that you simply store the `id` of your `Post` record as `post_id` and define a `post` method:

```ruby
def post
  Post.find(post_id)
end
```

**Comment by jasl** [comment]
You don't need to include`AM::Serialize`, AE and AllFutures already included.

I suggest don't dump AR object with Marshall, because AR object is complex... you can use `serializable_hash` or `to_h` that transform AR object to simple Hash, then dump it, or you can store the id and fetch it when you need.

**Comment by jasl** [comment]
BTW, if you wanna dump AR anyway, it's better to use YAML, see https://github.com/rails/rails/blob/19f8d03aba5f6ae15a7c9a89a39df71606fd9089/activerecord/lib/active_record/legacy_yaml_adapter.rb

and you shall load it via https://github.com/rails/rails/blob/main/activerecord/lib/active_record/core.rb#L471-L489

**Comment by Roupiye** [comment]
![image](https://user-images.githubusercontent.com/38011914/155410269-97727825-2e81-4899-84ea-b43301f82f74.png)
i fixed it saving the serialized post using kredis and storing the key on in post_id

the serializable_hash dont work with rich_text, and i would need to build my own serializer for posts since it has a lot of nested attributes and recursive nested attributes
![image](https://user-images.githubusercontent.com/38011914/155411274-02beebc0-7c30-4122-af4b-8a94d5d1a3bd.png)

**Comment by TopRoupi** [comment]
the older code would some times find a post with a post_id like "5dde148c-2887-4b10-8fdf-44b0c5e57bb8" it would find a post with id 5.
so only worked with a post_id that didn't started with number

 i changed it to:
![image](https://user-images.githubusercontent.com/38011914/155742298-73e4ca16-c329-4e55-8d25-8d676400535e.png)

**Comment by leastbad** [comment]
Breno, I saw your post from half an hour ago in my email. It seems like you have since deleted it? I am pretty confused... we want to help but you're not making it easy to do so.

As we have said previously... you should not be including `ActiveModel::Serialization` and we strongly recommend that you do not use `Marshall` on Active Record models. You should not need to do this... either use Yaml serialization or just use All Futures models to hold your unsaved Active Record model. I describe how this works in https://allfutures.leastbad.com/usage/active-record#creating-or-updating-from-an-active-record-model

I tried to follow what you were saying about UUID based ids that start with an integer, but I'm afraid I just don't follow. I need you to use more descriptive language; tell us what problem you're trying to solve and we'll tell you how to solve it. https://xyproblem.info/
